src/struc/model.jl

```diff
@@ -0,0 +1,826 @@
+    organisemodel,
+    organisestruc,
+    stdatomselector,
+    hetatomselector,
+    atomnameselector,
+    calpha_atom_names,
+    calphaselector,
+    backbone_atom_names,
+    backboneselector,
+    heavy_atom_names,
+    heavyatomselector,
+    resnameselector,
+    water_res_names,
+    waterselector,
+    stdresselector,
+    hetresselector,
+    disorderselector,
+    hydrogenselector
+
+
+import Base: getindex,
+    setindex!,
+    length,
+    start,
+    next,
+    done,
+    eltype,
+    show
+
+
+"""A protein structural element."""
+abstract StrucElement
+
+
+"""An atom represented in a PDB file - either an `Atom` or a
+`DisorderedAtom`."""
+abstract AbstractAtom <: StrucElement
+
+"""An atom record from a PDB file."""
+immutable Atom <: AbstractAtom
+    het_atom::Bool
+    serial::Int
+    name::AbstractString
+    alt_loc_id::Char
+    res_name::AbstractString
+    chain_id::Char
+    res_no::Int
+    ins_code::Char
+    coords::Array{Float64,1}
+    occupancy::Float64
+    temp_fac::Float64
+    element::AbstractString
+    charge::AbstractString
+end
+
+"""A container to hold different versions of the same atom."""
+immutable DisorderedAtom <: AbstractAtom
+    alt_loc_ids::Dict{Char, Atom}
+    default::Char
+end
+
+
+"""A residue (amino acid) or other molecule from a PDB file."""
+abstract AbstractResidue <: StrucElement
+
+"""A residue (amino acid) or other molecule from a PDB file."""
+immutable Residue <: AbstractResidue
+    name::AbstractString
+    chain_id::Char
+    number::Int
+    ins_code::Char
+    het_res::Bool # Does the residue consist of hetatoms?
+    atom_list::Array{AbstractString,1}
+    atoms::Dict{AbstractString, AbstractAtom}
+end
+
+Residue(name::AbstractString, chain_id::Char, number::Int, ins_code::Char, het_res::Bool) = Residue(name, chain_id, number, ins_code, het_res, AbstractString[], Dict())
+Residue(atom::AbstractAtom) = Residue(getresname(atom), getchainid(atom), getresnumber(atom), getinscode(atom), ishetatom(atom))
+
+"""A container to hold different versions of the same residue
+(point mutations)."""
+immutable DisorderedResidue <: AbstractResidue
+    names::Dict{AbstractString, Residue}
+    default::AbstractString
+end
+
+
+"""A chain from a PDB file."""
+immutable Chain <: StrucElement
+    id::Char
+    res_list::Array{AbstractString,1}
+    residues::Dict{AbstractString, AbstractResidue}
+end
+
+Chain(id::Char) = Chain(id, Dict())
+
+
+"""A model from a PDB file."""
+immutable Model <: StrucElement
+    number::Int
+    chains::Dict{Char, Chain}
+end
+
+Model(number::Int) = Model(number, Dict())
+
+
+"""A container for multiple models from a PDB file."""
+immutable Structure <: StrucElement
+    name::AbstractString
+    models::Dict{Int, Model}
+end
+
+Structure(name::AbstractString) = Structure(name, Dict())
+
+
+"""An `Array{AbstractAtom,1}`."""
+typealias AtomList Array{AbstractAtom,1}
+
+"""An `Array{AbstractResidue,1}`."""
+typealias ResidueList Array{AbstractResidue,1}
```

Comment:
These type aliases would be useless in many cases because `Vector{Atom}` objects are not an instance of `AtomList`. For example, `isa([1,2,3], Vector{Integer})` is `false` due to invariance of type parameters (http://julia.readthedocs.org/en/latest/manual/types/#parametric-composite-types). In general, using abstract types as type parameters should be avoided.